File: built-in plugins dsp/mb-piano-grand.cpp
/**
 * MB Grand Piano
 * Category : instrument
 * Type     : piano
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Concert grand piano with rich harmonics
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PIANO_GRAND_H
#define MB_PIANO_GRAND_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPianoGrand : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-piano-grand";
    static constexpr const char* PLUGIN_NAME    = "MB Grand Piano";
    static constexpr const char* PLUGIN_TYPE    = "piano";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float brightness = 0.5f;  // range [0, 1]
    float dynamics = 0.7f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbPianoGrand() = default;
    ~MbPianoGrand() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.dynamics = std::clamp(params.dynamics, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Grand Piano
        return input;
    }
};

#endif // MB_PIANO_GRAND_H

File: built-in plugins dsp/mb-piano-grand.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbPianoGrandPlugin: PluginDefinition = {
  id: "mb-piano-grand",
  slug: "mb-piano-grand",
  name: "MB Grand Piano",
  category: "instrument",
  type: "piano",
  version: "1.0.0",
  description: "Concert grand piano with rich harmonics",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "triangle", detune: 0, gain: 0.6 },
    { type: "sine", detune: 0.5, gain: 0.3 },
    { type: "sine", detune: 1200, gain: 0.1 },
  ],
  envelope: { attack: 0.002, decay: 0.3, sustain: 0.6, release: 0.5 },
  parameters: [
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "dynamics",
      name: "Dynamics",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { brightness: 0.5, dynamics: 0.7, volume: 0.8 },
};

export default MbPianoGrandPlugin;
